src/cpp/session/SessionSuspend.cpp

```diff
@@ -41,15 +46,222 @@ bool s_suspendedFromTimeout = false;
 // was the underlying r session resumed
 bool s_rSessionResumed = false;
 
+// keep track of what operations are blocking session from suspending
+enum SuspendTimeoutState
+{
+   kWaitingForTimeout,
+   kWaitingForNonBlocking
+};
+
+std::vector<bool> opsBlockingSuspend(kBlockingOpsCount, false);
+boost::posix_time::ptime s_suspendTimeoutTime = boost::posix_time::second_clock::universal_time();
+bool s_dirtyBlockingOps = false;
+SuspendTimeoutState s_timeoutState = kWaitingForTimeout;
+
 } // anonymous namespace
 
 // convenience function for disallowing suspend (note still doesn't override
 // the presence of s_forceSuspend = 1)
-bool disallowSuspend() 
-{ 
+bool disallowSuspend()
+{
    return false;
 }
 
+boost::posix_time::ptime timeoutTimeFromNow()
+{
+   int timeoutMinutes = options().timeoutMinutes();
+   if (timeoutMinutes > 0)
+   {
+      return boost::posix_time::second_clock::universal_time() +
+             boost::posix_time::minutes(options().timeoutMinutes());
+   }
+   else
+   {
+      return boost::posix_time::ptime(boost::posix_time::not_a_date_time);
+   }
+}
+
+bool isTimedOut(const boost::posix_time::ptime& timeoutTime)
+{
+   using namespace boost::posix_time;
+
+   // never time out in desktop mode
+   if (options().programMode() == kSessionProgramModeDesktop)
+      return false;
+
+   // check for an client disconnection based timeout
+   int disconnectedTimeoutMinutes = options().disconnectedTimeoutMinutes();
+   if (disconnectedTimeoutMinutes > 0)
+   {
+      ptime lastEventConnection =
+         httpConnectionListener().eventsConnectionQueue().lastConnectionTime();
+      if (!lastEventConnection.is_not_a_date_time())
+      {
+         if ( (lastEventConnection + minutes(disconnectedTimeoutMinutes)
+               < second_clock::universal_time()) )
+         {
+            return true;
+         }
+      }
+   }
+
+   // check for a foreground inactivity based timeout
+   if (timeoutTime.is_not_a_date_time())
+      return false;
+   else
+      return second_clock::universal_time() > timeoutTime;
+}
+
+void resetSuspendTimeout()
+{
+   s_suspendTimeoutTime = timeoutTimeFromNow();
+}
+
+core::json::Object blockingOpsToJson()
+{
+   core::json::Object blockingOps;
```

Comment:
Nit: Rather than name this the reverse order of the other array, I'd name it almost the same thing but with `Json` at the end, as it more clearly explains their difference